File: Lv 3/디스크 컨트롤러.py
import heapq

def solution(jobs):
    answer = 0
    heap = []
    
    jobs.sort()
    currentTime = jobs[0][0]; # 초기 작업의 시간으로 세팅
    jobidx = 1
    
    # 작업 시간을 기준으로 최소 히프 구성
    heapq.heappush(heap, [jobs[0][1], jobs[0][0]])
    
    while heap:
        # 현재 기준 최소 시간의 작업 가져옴
        jobtime, arrive = heapq.heappop(heap)
        currentTime += jobtime
        while jobidx < len(jobs):
            if jobs[jobidx][0] <= currentTime:
                heapq.heappush(heap, [jobs[jobidx][1], jobs[jobidx][0]])
                jobidx += 1
            else:
                break

        answer += (currentTime - arrive)
        
        if not heap and jobidx < len(jobs):
            currentTime = jobs[jobidx][0]
            heapq.heappush(heap, [jobs[jobidx][1], jobs[jobidx][0]])
            jobidx += 1

    
    answer = answer // len(jobs)
    return answer

File: Lv 3/디스크 컨트롤러.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.PriorityQueue;

class Job implements Comparable<Job> {

    int arriveTime;
    int duration;

    public Job(int arriveTime, int duration) {
        this.arriveTime = arriveTime;
        this.duration = duration;
    }

    @Override
    public int compareTo(Job o) {
        if (this.duration > o.duration) {
            return 1;
        }
        if (this.duration == o.duration) {
            return -Integer.compare(this.arriveTime, o.arriveTime);
        }
        return -1;
    }

    @Override
    public String toString() {
        return "Job{" +
                "arriveTime=" + arriveTime +
                ", duration=" + duration +
                '}';
    }
}

class Solution {

    public int solution(int[][] jobs) {
        int jobIndex = 0;
        int totalTime = 0;
        int endTime = 0;

        Arrays.sort(jobs, Comparator.comparingInt(o -> o[0]));

        PriorityQueue<Job> heap = new PriorityQueue<>();

        for (int time = 0; time <= 1000001; time++) {
            while (jobIndex < jobs.length && jobs[jobIndex][0] <= time) {
                heap.offer(new Job(jobs[jobIndex][0], jobs[jobIndex][1]));
                jobIndex++;
            }
            if (time >= endTime && !heap.isEmpty()) {
                Job job = heap.poll();
                endTime = time + job.duration;
                totalTime += (time + job.duration) - job.arriveTime;
            }
        }

        return totalTime / jobs.length;
    }
}
